components/Hooks/useLocation.jsx

```diff
@@ -0,0 +1,38 @@
+import { useState, useEffect } from "react";
+const BASE_URL = `https://api.tfl.gov.uk/`;
+export default function useLocation(options) {
+  const [lat, setLat] = useState(0);
+  const [lon, setLon] = useState(0);
+  const [station, setStation] = useState(null);
+
+  // when user click the pin then get user's location.
+  useEffect(() => {
+    if (lat !== 0 && lon !== 0) {
+      fetch(
+        `${BASE_URL}StopPoint?lat=${lat}&lon=${lon}&stopTypes=NaptanMetroStation`
+      )
+        .then((response) => {
+          if (response.ok) {
+            return response.json();
+          }
+        })
+        .then((result) => {
+          if (result) {
+            //at space 4, user will get finsbury park as staiton naem.
+            setStation(
+              result.stopPoints[0].commonName.replace("Underground Station", "")
+            );
+            // add user location's station inside options as first row.
+            if (station !== null) {
+              options.unshift({
+                value: `${lat},${lon}`,
+                label: `📍 Your Location`,
+              });
+            }
```

Comment:
There's at least two approaches we can take to the UX which will affect what needs to happen here.

Either we can ask TfL's API for the nearest station based on GPS ("find my nearest station"), or we can tell the form that the user wants to use their location and have TfL's journey planner find the best starting station for a selected end-station ("use my location").

If the former, I think a better approach would be something like this:

- Check if the station that was found from the GPS exists in our list of stations (``options``).
- If so, tell the react-select component to select the station that was found (rather than adding a new item to the list).

If the latter, the "Your location" item should get selected in the react-select list. (At the moment it gets added, but not selected). (We'll also have to answer the UX question of how to tell the user that they need to get to the suggested starting station under their own power.)

Neither of those things need fixing here and now, I just wanted to make a note so we don't forget them later! :)

Reply:
TfL's API for the nearest station based on GPS ("find my nearest station") any links for this api?

Reply:
I think that's what the code in there is already doing? (``${BASE_URL}StopPoint?lat=${lat}&lon=${lon}&stopTypes=NaptanMetroStation`` etc.)